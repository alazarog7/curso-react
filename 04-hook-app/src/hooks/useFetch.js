import React, { useEffect, useRef, useState } from 'react'

export const useFetch = (url) => {

    const isMounted = useRef(true);
    const [state, setstate] = useState({data: null, loading: true, error: null});

    useEffect(() => {
        return () => {
            isMounted.current = false;
        }
    }, []);

    useEffect(() => {
        fetch(url).then( resp => {
                                    return resp.json()
                                 })
                  .then(data => {
                                    //setTimeout(() => {
                                        if(isMounted.current){
                                            setstate({
                                                loading : false,
                                                error: null, 
                                                data
                                            })
                                        } else {
                                            console.log("Se interrumpio");
                                        }
                                    //}, 3000);
                                })
                  .catch(()=>{
                      setstate({
                          data:null,
                          loading: false,
                          error:'No se cargo la data' 
                      })
                  })
    }, [url]);
    return state;
}
