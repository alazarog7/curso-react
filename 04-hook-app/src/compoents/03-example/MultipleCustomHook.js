import React from 'react'
import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch'

export const MultipleCustomHook = () => {

    const {state, increment } = useCounter(1)
    const {loading, data, error} = useFetch(`https://www.breakingbadapi.com/api/quotes/${state}`);
    const {author,quote} = !!data&&data[0];

    
    return (
        <div>
            <h1>Breaking Quotes</h1>
            <hr/>
            {
                loading ? 
                (
                    <div className="alert alert-info text-center">Loading ... </div>
                ): 
                (
                        <blockquote className="blockquote text-right">
                          <p className="mb-0">{quote}</p>
                          <footer className="card-blockquote">{author}</footer>
                        </blockquote>
                )

            }

            <button className="btn btn-primary" onClick={()=>increment()}>
                Siguiente Quote
            </button>
        </div>
    )
}
