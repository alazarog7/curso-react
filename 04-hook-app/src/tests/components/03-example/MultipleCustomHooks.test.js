import shallow from "enzyme/build/shallow"
import { React } from 'react';
import { MultipleCustomHook } from '../../../compoents/03-example/MultipleCustomHook';
import { useFetch } from "../../../hooks/useFetch";
jest.mock('../../../hooks/useFetch');

describe('Multiple Custom Hooks Set Tests', () => {
    test('should ', () => {
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });
        const wrapper = shallow(<MultipleCustomHook/>)
        expect(wrapper).toMatchSnapshot();
    })

    test('should return the right data ', () => {
        
        useFetch.mockReturnValue({
            data: [
                {
                    author: 'Alejandro',
                    quote: 'El error existe' 
                }
            ],
            loading: false,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHook/>)

        expect(wrapper.find('.alert').exists()).toBe(false);
        expect(wrapper.find('.mb-0').text().trim()).toBe('El error existe');
        expect(wrapper.find('footer').text().trim()).toBe('Alejandro');



    })
    
    
})
