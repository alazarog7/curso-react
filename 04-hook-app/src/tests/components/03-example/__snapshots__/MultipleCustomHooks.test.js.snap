// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multiple Custom Hooks Set Tests should  1`] = `
<div>
  <h1>
    Breaking Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading ... 
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente Quote
  </button>
</div>
`;
